x-pack/plugins/endpoint/server/routes/resolver/utils/tree.ts

```diff
@@ -0,0 +1,243 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import _ from 'lodash';
+import { ResolverEvent } from '../../../../common/types';
+import { extractEntityID, extractParentEntityID } from './normalize';
+import { buildPaginationCursor } from './pagination';
+
+interface NodeStats {
+  totalEvents: number;
+  totalAlerts: number;
+}
+
+interface NodePagination {
+  nextChild?: string | null;
+  nextEvent?: string | null;
+  nextAncestor?: string | null;
+  nextAlert?: string | null;
+}
+
+interface Node {
```

Comment:
Probably worth a comment here that even though Node has both children and parent fields, they aren't used at the same time so we don't create a fully bidirectional tree structure - instead our root is in the middle of the tree with a line of parents above and then branching out with children below.  This makes sense for our case but isn't the normal tree/root structure.